src/hotspot/share/utilities/waitBarrier_generic.cpp

```diff
@@ -29,66 +30,229 @@
-    OrderAccess::loadload();
-    return 0;
+void GenericWaitBarrier::Cell::arm(int32_t requested_tag) {
+  // Before we continue to arm, we need to make sure that all threads
+  // have left the previous cell.
+
+  int64_t state;
+
+  SpinYield sp;
+  while (true) {
+    state = Atomic::load_acquire(&_state);
+    assert(decode_tag(state) == 0,
+           "Pre arm: Should not be armed. "
+           "Tag: " INT32_FORMAT "; Waiters: " INT32_FORMAT,
+           decode_tag(state), decode_waiters(state));
+    if (decode_waiters(state) == 0) {
+      break;
+    }
+    sp.wait();
   }
-  assert(w > 0, "Bad counting");
-  // We need an exact count which never goes below zero,
-  // otherwise the semaphore may be signalled too many times.
-  if (Atomic::cmpxchg(&_waiters, w, w - 1) == w) {
-    _sem_barrier.signal();
-    return w - 1;
+
+  // Try to swing cell to armed. This should always succeed after the check above.
+  int64_t new_state = encode(requested_tag, 0);
+  int64_t prev_state = Atomic::cmpxchg(&_state, state, new_state);
+  if (prev_state != state) {
+    fatal("Cannot arm the wait barrier. "
+          "Tag: " INT32_FORMAT "; Waiters: " INT32_FORMAT,
+          decode_tag(prev_state), decode_waiters(prev_state));
   }
-  return w;
 }
 
-void GenericWaitBarrier::disarm() {
-  assert(_barrier_tag != 0, "Not armed");
-  _barrier_tag = 0;
-  // Loads of _barrier_threads/_waiters must not float above disarm store and
-  // disarm store must not sink below.
-  OrderAccess::fence();
-  int left;
+int GenericWaitBarrier::Cell::signal_if_needed(int max) {
+  int signals = 0;
+  while (true) {
+    int cur = Atomic::load_acquire(&_outstanding_wakeups);
+    if (cur == 0) {
+      // All done, no more waiters.
+      return 0;
+    }
+    assert(cur > 0, "Sanity");
+
+    int prev = Atomic::cmpxchg(&_outstanding_wakeups, cur, cur - 1);
+    if (prev != cur) {
+      // Contention, return to caller for early return or backoff.
+      return prev;
+    }
+
+    // Signal!
+    _sem.signal();
+
+    if (++signals >= max) {
+      // Signalled requested number of times, break out.
+      return prev;
+    }
+  }
+}
+
+void GenericWaitBarrier::Cell::disarm(int32_t expected_tag) {
+  int32_t waiters;
+
   SpinYield sp;
-  do {
-    left = GenericWaitBarrier::wake_if_needed();
-    if (left == 0 && _barrier_threads > 0) {
-      // There is no thread to wake but we still have barrier threads.
+  while (true) {
+    int64_t state = Atomic::load_acquire(&_state);
+    int32_t tag = decode_tag(state);
+    waiters = decode_waiters(state);
+
+    assert((tag == expected_tag) && (waiters >= 0),
+           "Mid disarm: Should be armed with expected tag and have sane waiters. "
+           "Tag: " INT32_FORMAT "; Waiters: " INT32_FORMAT,
+           tag, waiters);
+
+    int64_t new_state = encode(0, waiters);
+    if (Atomic::cmpxchg(&_state, state, new_state) == state) {
+      // Successfully disarmed.
+      break;
+    }
+    sp.wait();
```

Comment:
I don't think it is hurting, and some platform do not have a CAS, so Atomic::cmpxchg may be a load-reserve conditional store (risc-v). Which have the possibility of all failed with the cmpxchg.

These are the default, so we do spin 4096 times before we yield. (then yields 64 times before it sleeps)
```
  static const uint default_spin_limit = 4096;
  static const uint default_yield_limit = 64;
  static const uint default_sleep_ns = 1000;
```

Reply:
https://lore.kernel.org/all/20230910082911.3378782-10-guoren@kernel.org/

Reply:
Actually, I think we don't need `SpinYield` in this particular place for a few reasons:
 1. We want to disarm as fast as possible, even if that means more contention, since we are on "leaving safepoint" path in VM thread here.
 2. No other `_state` update loop yields, so this loop is effectively low priority under contention, which makes (1) even worse.
 3. There is a sharing between `SpinYield` in `_state` CAS loop here and wakeup backoff later. Which is subtly leaking the `SpinYield` state between the phases: the aggressive backoff accrued due to `_state` contention would transfer to dealing with signaling contention.

I removed this `wait()` in new commit.

Reply:
Yes, you are correct, since they are moving towards identical state, thanks!